File: day3/py/main.py
from time import time

def isDig(ch):
    return (ch > 47) and (ch < 58)

def isSym(ch):
    return ((ch > 32) and (ch < 46)) or ((ch > 57) and (ch < 65)) or ch == 47

def isGear(ch):
    return ch == 42

class MyInput:
    def __init__(self, lines):
        self.barray = bytearray()
        self.span = len(lines[0]) + 2
        for line in lines:
            self.barray.extend(b'.')
            self.barray.extend(map(ord, line))
            self.barray.extend(b'.')
        self.barray.extend(map(ord, self.span * '.'))

    def __getitem__(self, index):
        return self.barray[index]

    def __len__(self):
        return len(self.barray)

    def __getxy(self, x, y):
        return self.barray[y*self.span + x]

    def checkSym(self, index):
        y = index // self.span
        x = index % self.span
        for comb in [(0, -1), (1, -1), (1, 0), (1, 1), (0, 1), (-1, 1), (-1, 0), (-1, -1)]:
            if (isSym(self.__getxy(x + comb[0], y + comb[1]))):
                return True
        return False

    def checkGear(self, index):
        y = index // self.span
        x = index % self.span
        for comb in [(0, -1), (1, -1), (1, 0), (1, 1), (0, 1), (-1, 1), (-1, 0), (-1, -1)]:
            if (isGear(self.__getxy(x + comb[0], y + comb[1]))):
                return (y + comb[1]) * self.span + x + comb[0]
        return -1

def checkAndSum(input):
    nums = []
    numstr = []
    flag = False
    for i in range(len(input)):
        if isDig(input[i]):
            numstr.append(chr(input[i]))
            flag = True if flag else input.checkSym(i)
        elif len(numstr) > 0 and flag:
            nums.append(int(''.join(numstr)))
            numstr.clear()
            flag = False
        elif len(numstr) > 0:
            numstr.clear()
    
    return sum(nums)

def appendKey(dict, key, val):
    if key not in dict:
        dict[key] = list()
    dict[key].append(val)
    return dict

def findGear(input):
    gears = {}
    numstr = []
    gearid = -1
    for i in range(len(input)):
        if isDig(input[i]):
            numstr.append(chr(input[i]))
            gearid = gearid if (gearid >= 0) else input.checkGear(i)
        elif len(numstr) > 0 and gearid >= 0:
            gears = appendKey(gears, gearid, int(''.join(numstr)))
            numstr.clear()
            gearid = -1
        elif len(numstr) > 0:
            numstr.clear()

    res = 0
    for value in gears.values():
        if len(value) == 2:
            res += value[0] * value[1]
    return res

input = \
"""467..114..
...*......
..35..633.
......#...
617*......
.....+.58.
..592.....
......755.
...$.*....
.664.598.."""
lines = open("../input.txt", "r").readlines()
myinput = MyInput(lines)

#Part One
start = time()
print("part one =", checkAndSum(myinput))
ms = round((time() - start) * 1000)
print("time[ms] =", ms)

#Part Two
start = time()
print("part two =", findGear(myinput))
ms = round((time() - start) * 1000)
print("time[ms] =", ms)

File: day3/js/main.js
function isDigit(c) {
    return c >= '0' && c <= '9';
}

function isSym(c) {
    return (c >= '!' && c <= '-') || c == '/' || (c >= ':' && c <= '@');
}

function prepInput(lines) {
    const span = lines[0].length + 1;
    const newlines = [];
    newlines.push('.'.repeat(span));
    lines.map((line) => newlines.push('.' + line + '.'));
    newlines.push('.'.repeat(span));
    return newlines;
}

function findSym(lines, row, col, ndigs) {
    const near = lines[row-1].substr(col-1, ndigs+2) +
    lines[row+1].substr(col-1, ndigs+2) +
    lines[row][col-1] + lines[row][col+ndigs];

    for (let c of near) {
        if (isSym(c)) return true;
    }
    return false
}

function findGear(lines, row, col, ndigs) {
    const near = []
    near.push(lines[row-1].substr(col-1, ndigs+2));
    near.push(lines[row].substr(col-1, ndigs+2));
    near.push(lines[row+1].substr(col-1, ndigs+2));
    for (let r = 0; r < near.length; r++) {
        for (let c = 0; c < near[r].length; c++) {
            if (near[r][c] == '*') {
                return (row-1+r).toString() + ',' + (col-1+c).toString();
            }
        }
    }
    return null;
}

function partOne(lines) {
    const nums = [];
    for (let row = 0; row < lines.length; row++) {
        for (let col = 0; col < lines[row].length; col++) {
            if (isDigit(lines[row][col])) {
                const num = parseInt(lines[row].substr(col, 3))
                const ndigs = num.toString().length
                if (findSym(lines, row, col, ndigs)) {
                    nums.push(num)
                }
                col += ndigs
            }
        }
    }

    let sum = 0
    nums.forEach( num => {
        sum += num
    });
    return sum
}

function partTwo(lines) {
    const gears = {}
    for (let row = 0; row < lines.length; row++) {
        for (let col = 0; col < lines[row].length; col++) {
            if (isDigit(lines[row][col])) {
                const num = parseInt(lines[row].substr(col, 3));
                const ndigs = num.toString().length;
                const gearid = findGear(lines, row, col, ndigs);
                if (gearid) {
                    if (gearid in gears) {
                        gears[gearid].push(num)
                    } else {
                        gears[gearid] = [];
                        gears[gearid].push(num)
                    }
                }
                col += ndigs;
            }
        }
    }

    let sum = 0
    Object.entries(gears).forEach(
        ([key, val]) => { 
            if (val.length == 2) {
                sum += val[0] * val[1];
            }
    });
    return sum;
}

const fs = require('fs');
fs.readFile('../input.txt', 'utf8', function (err, data) {
    if (err) {
        console.error(err);
        return;
    }
    var lines = prepInput(data.split('\n'));

    const start1 = performance.now();
    const res1 = partOne(lines);
    const end1 = performance.now();
    console.log("part one =", res1);
    console.log("time[ms] =", end1 - start1);

    const start2 = performance.now();
    const res2 = partTwo(lines)
    const end2 = performance.now();
    console.log("part two =", res2);
    console.log("time[ms] =", end2 - start2);
});